src/newButton/Button.tsx

```diff
@@ -0,0 +1,234 @@
+      })
+    : icon;
+};
+
+export const Button = forwardRefWithAs<
+  ButtonProps,
+  HTMLButtonElement,
+  "button"
+>((props, ref) => {
+  const {
+    children,
+    size = "sm",
+    variant = "solid",
+    iconOnly,
+    suffix,
+    prefix,
+    loading = false,
+    spinner,
+    disabled = false,
+    className,
+    ...rest
+  } = props;
+  const _disabled = disabled || loading;
+
+  const theme = useTheme();
+  const baseStyles = cx(
+    theme.newButton.base,
+    !iconOnly
+      ? theme.newButton.size[size]
+      : theme.newButton.iconOnly.size[size],
+    theme.newButton.variant[variant],
+    className,
+  );
+
+  return (
+    <ReakitButton
+      ref={ref}
+      className={baseStyles}
+      disabled={_disabled}
+      {...rest}
+    >
+      {loading && !suffix && !prefix ? (
+        <>
+          <ButtonSpinnerWrapper>
+            <ButtonSpinner spinner={spinner} iconOnly={iconOnly} size={size} />
+          </ButtonSpinnerWrapper>
+          <div className="opacity-0">
+            <ButtonChildren
+              iconOnly={iconOnly}
+              suffix={suffix}
+              prefix={prefix}
+              size={size}
+              loading={loading}
+              spinner={spinner}
+            >
+              {children}
+            </ButtonChildren>
+          </div>
+        </>
+      ) : (
+        <ButtonChildren
+          iconOnly={iconOnly}
+          suffix={suffix}
+          prefix={prefix}
+          size={size}
+          loading={loading}
+          spinner={spinner}
+        >
+          {children}
+        </ButtonChildren>
+      )}
+    </ReakitButton>
+  );
+});
+
+Button.displayName = "Button";
+
+type ChildrenWithPrefixSuffixProps = {
+  suffix?: React.ReactElement;
+  prefix?: React.ReactElement;
+  spinner?: React.ReactElement;
+  size: NonNullable<ButtonProps["size"]>;
+  loading?: boolean;
+};
+
+const ChildrenWithPrefixSuffix: React.FC<ChildrenWithPrefixSuffixProps> =
+  props => {
+    const { suffix, prefix, children, size, loading, spinner } = props;
+    const theme = useTheme();
+    const suffixStyles = cx(theme.newButton.suffix.size[size]);
+    const prefixStyles = cx(theme.newButton.prefix.size[size]);
+
+    return (
+      <>
+        {prefix ? (
+          loading && !suffix ? (
+            <ButtonSpinner spinner={spinner} size={size} />
+          ) : (
+            runIfFn(withIconA11y(prefix, { className: prefixStyles }))
+          )
+        ) : null}
+        <span>{children}</span>
+        {suffix ? (
+          loading ? (
+            <ButtonSpinner spinner={spinner} size={size} />
+          ) : (
+            runIfFn(withIconA11y(suffix, { className: suffixStyles }))
+          )
+        ) : null}
+      </>
+    );
+  };
+
+type ButtonChildrenProps = {
+  iconOnly?: React.ReactElement;
+  spinner?: React.ReactElement;
+  suffix?: React.ReactElement;
+  prefix?: React.ReactElement;
+  size: NonNullable<ButtonProps["size"]>;
+  loading?: boolean;
+};
```

Comment:
Create a common interface to share between all the components

```tsx
interface ButtonChildrenCommonProps {
  suffix?: React.ReactElement;
  prefix?: React.ReactElement;
  spinner?: React.ReactElement;
  size: NonNullable<ButtonProps["size"]>;
  loading?: boolean;
}

interface ChildrenWithPrefixSuffixProps extends ButtonChildrenCommonProps {}


interface ButtonChildrenProps extends ButtonChildrenCommonProps {
  iconOnly?: React.ReactElement;
}


interface ButtonSpinnerProps
  extends Pick<ButtonChildrenCommonProps, "spinner" | "size"> {
  iconOnly?: React.ReactElement;
}

```

Reply:
Oh we pick from ButtonProps too 

```ts
interface ButtonChildrenCommonProps
  extends Pick<
    ButtonProps,
    "suffix" | "prefix" | "spinner" | "size" | "loading"
  > {
  size: NonNullable<ButtonProps["size"]>;
}
```

Reply:
Yeah, right we can pick them up from Button Props.
I did for the size, but did not strike me to do for the others.